MOxUnit/assertExceptionThrown.m

```diff
@@ -1,63 +1,120 @@
 %
-% Throws:
-%   'moxunit:exceptionNotRaised'    func() does not raise an exception
-%   'moxunit:wrongExceptionRaised'  func() does raise an exception but with
-%                                       an identifier different from
-%                                       expected_id
+%   % Assert that sin throws AND that the ID is 'MATLAB:UndefinedFunction'
+%   >> assertExceptionThrown( @()sin( struct([]) ), 'MATLAB:UndefinedFunction')
+%  
+% INPUT:
+%   func        - Function handle that is expected to throw, with the prototype
+%                   [varargout{:}] = func()
+%   expectedID  - Idenitifier of the expected exception.  If expectedID is
+%                 empty, then any exception (identifier) is allowed.  When 
+%                 exactly two inputs are passed an ambiguity arises between
+%                 expectedID and message because these inputs share a common
+%                 type.  By default, we resolve this ambiguity using
+%                 moxunit_util_is_message_identifier().  If any exception is
+%                 permitted, we recommend explicitly defining expectedID to be
+%                 empty.
+%                   Default: ''
+%   message     - Custom message to be included when func fails to throw
+%   
+% THROWS:
+%   'moxunit:exceptionNotRaised'    - func() does not raise an exception but was
+%                                     expected to do so.
+%   'moxunit:wrongExceptionRaised'  - func() does raise an exception but with
+%                                     an identifier different from expected_id
 %
+% ASSUMPTIONS: 
+%   All input variables are of the correct type, valid (if applicable), and 
+%   given in the correct order. 
 %
-% Raises:
-%
-    exception_was_raised=false;
-    if moxunit_util_platform_is_octave()
-        try
-            func();
-        catch
-            exception_was_raised=true;
-            [unused,found_id]=lasterr();
-        end
-    else
-        try
-            func();
-        catch mexception;
-            exception_was_raised=true;
-            found_id=mexception.identifier;
+% See also moxunit_util_is_message_identifier
+%-------------------------------------------------------------------------------
+%}
+function assertExceptionThrown(func, expectedID, message)
+
+% Parse Inputs
+switch nargin
+    case 1
+        expectedID = '';
+        message = '';
+        
+    case 2
+        if moxunit_util_is_message_identifier(expectedID)
+            message = '';
+        else
+            message = expectedID;
+            expectedID = '';
         end
+        
+    case 3
+        % Do nothing
+        
+    otherwise
+        error('assertExceptionThrown:unexpectedInput',...
+            'Unexpected number of inputs');
+end
+
+
+% Check func for an exception and capture it
+funcException = false;
+if moxunit_util_platform_is_octave()
+    try
+        func();
+    catch
+        funcException = true;
+        [~,foundID] = lasterr();
```

Comment:
The tilde (``~``) is not supported in early versions of Matlab. Can we use a real variable, such as ``unused``, instead?